### ISSUE TITLE
Cluster support with Kitchen

### ISSUE THREAD

**Issue by Lax77** [test-kitchen/test-kitchen#905]
I have a use case where I need to spin off a cluster (A Master and multiple Slaves) and run test suites against them. 
 - Would like Master node aware of Slave nodes and vice versa 
 - Carry out regular checks on master instance to ensure all the expected SW components are installed and running
 - Similarly on Slave nodes check expected software components are installed and running

So far I could get both Master & Slave deployed on a single Instance with kitchen and run my test suites. However I am interested to have Kitchen launch separate instance for Master & Slave and run test suites. Is this possible with Kitchen? Any pointers are appreciated.

**Comment by cheeseplus** [comment]
tl;dr This is not possible with test-kitchen currently. 

Longer version:
This has been something folks have been punting around for at least 2 years in the community but at present is still very much out of scope for TK. It's something a lot of us think would be neat but the TK design has always been targeted at doing one thing and doing it well. If we do take it on, it will be after paying off a bunch of technical debt before adding any more features - ideally this would be something we could abstract to a driver and keep out of core TK if possible.

**Comment by mwrock** [comment]
Just to throw out how you might accomplish this today is to use the [kitchen-nodes](https://github.com/mwrock/kitchen-nodes) plugin that will allow nodes to use chef search and find each other in kitchen runs.

The clunky thing is you do need to create/converge the nodes separately. You can orchestrate this in a rakefile. Something like this:
```
  desc "run integration tests"
  task :integration do
    system('kitchen destroy')
    system('kitchen converge primary-windows-2012R2')
    system("kitchen exec primary-windows-2012R2 -c 'Restart-Computer -Force'")
    system('kitchen converge backup-windows-2012R2')
    system("kitchen exec backup-windows-2012R2 -c 'Restart-Computer -Force'")
    system('kitchen verify primary-windows-2012R2')
    system('kitchen verify backup-windows-2012R2')
  end
```
Clunky but it works.

I blogged about doing this with kitchen/docker/couchbase [here](http://www.hurryupandwait.io/blog/multi-node-testing-with-test-kitchen-and-docker).

**Comment by Lax77** [comment]
Thanks for the responses. Still wish it was supported with TK. Are there any other tool, you guys aware of will help in this space outside TK?

Matt, will try out the kitchen-nodes. Thanks for the pointer to the blog.

**Comment by cheeseplus** [comment]
We all, at various points, wish it was supported as well but it's a drastic divergence from our core competence. We'd love to add it but the effort/dedication required is significant not to mention there is plenty of debt that needs to be paid off first. There really aren't any great alternatives that I am aware of but the related projects may be helpful even if dated.

Related projects: 
https://github.com/chrisroberts/vagabond - tried to do this by wrapping TK but is kinda niche as it requires Librarian (vs Berkshelf) also not touched for a bit
https://github.com/Atalanta/leibniz - author left the project and it's been kinda unmaintained

**Comment by Lax77** [comment]
Ok, Thanks for the pointers Seth

**Comment by cheeseplus** [comment]
Closing this in favor of https://github.com/test-kitchen/test-kitchen/issues/873

**Comment by neillturner** [comment]
i'm doing this with kitchen-ansible but that's because ansible support remote ssh. see http://ec2dream.blogspot.co.uk/2015/12/automated-devops-with-test-kitchen.html
to me if chef could run in a remote ssh fashion like ansible then this would be easy.

**Comment by Lax77** [comment]
Interesting way to connect, thanks for sharing.

To get around the cluster limitation, I ended up creating 2 flows in my
case.

1. Single Node Testing
 - Use Kitchen + Ansible Spec
 - Fire up single instance
 - Provision infra (Master and Slave both)
 - Fire up the test suite

2. For cluster testing ( Vagrant + Ansible for provision, Serverspec + Rake
for firing test suites)
 - Have Vagrant to provision cluster
    - Which means fire up cluster(multiple masters and slave instances) on
env of choice (Virtual (Virtual Box) / Cloud (EC2, Open Stack, RackSpace,
etc))
    - Provision infra with Ansible on each of the instance that vagrant
provisioned
- Generate Inventory file(either manually or dynamically) for my cluster,
basically detailing IP and server roles(Master/Slave) etc.
- Use Serverspec + Rake to connect to the provisioned cluster and fire up
test suites

Also as I didn't want to maintain separate test cases for both the flows, I
ended up tweaking Rake definition to look up the test suites that were
written for Kitchen.

**Comment by retr0h** [comment]
For you ansible fans, check out [molecule](http://molecule.readthedocs.io).  This solves your particular issues with test-kitchen.
